compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -1532,11 +1532,85 @@ private boolean checkSimilarMatchGuard(BLangMatchGuard firstMatchGuard, BLangMat
             BLangTypeTestExpr secondTypeTest = (BLangTypeTestExpr) secondMatchGuard.expr;
             return ((BLangSimpleVarRef) firstTypeTest.expr).variableName.toString().equals(
                     ((BLangSimpleVarRef) secondTypeTest.expr).variableName.toString()) &&
-                    firstTypeTest.typeNode.getBType().tag == secondTypeTest.typeNode.getBType().tag;
+                    checkSimilarTypeDescInTypeTest(firstTypeTest.typeNode.getBType(),
+                            secondTypeTest.typeNode.getBType());
         }
         return false;
     }
 
+    private boolean checkSimilarTypeDescInTypeTest(BType firstType, BType secondType) {
+        if (firstType.tag != secondType.tag) {
+            return false;
+        }
+        switch (firstType.tag) {
+            case TypeTags.FINITE:
+                BFiniteType firstFiniteType = (BFiniteType) firstType;
+                BFiniteType secondFiniteType = (BFiniteType) secondType;
+                if (firstFiniteType.getValueSpace().size() != secondFiniteType.getValueSpace().size()) {
+                    return false;
+                }
+                for (BLangExpression expr1 : firstFiniteType.getValueSpace()) {
+                    boolean exprFound = false;
+                    for (BLangExpression expr2 : secondFiniteType.getValueSpace()) {
+                        if (expr2.getKind() == NodeKind.LITERAL && expr1.getKind() == NodeKind.LITERAL) {
+                            exprFound = ((BLangLiteral) expr2).value.equals(((BLangLiteral) expr1).value);
+                        } else if (expr2.getKind() == NodeKind.NUMERIC_LITERAL &&
+                                expr1.getKind() == NodeKind.NUMERIC_LITERAL) {
+                            exprFound = ((BLangNumericLiteral) expr2).value
+                                    .equals(((BLangNumericLiteral) expr1).value);
```

Comment:
With the following change, we can avoid some unnecessary checks right?

```suggestion
                        if (expr2.getKind() == NodeKind.LITERAL) {
                            exprFound = expr1.getKind() == NodeKind.LITERAL && ((BLangLiteral) expr2).value.equals(((BLangLiteral) expr1).value);
                        } else if (expr2.getKind() == NodeKind.NUMERIC_LITERAL) {
                            exprFound = expr1.getKind() == NodeKind.NUMERIC_LITERAL && ((BLangNumericLiteral) expr2).value
                                    .equals(((BLangNumericLiteral) expr1).value);
```

Reply:
Can probably extract the kinds out to variables too.